### PR TITLE
fix(deisctl): Switch from channels to io.Writers.

### PR DESCRIPTION
**Pull request by technosophos**
Fixes issue #3971 

deisctl used channels to send output and errors to a goroutine which
then wrote them. This patch replaces that with simpler code that simply
writes messages to Stdout and Stderr (configurable for outher writers).

### CODE REVIEW

**Comment by technosophos** [comment]
I "fixed" the line overwriting behavior in this release, but I'm not sure that we should keep it. It seems to me that we are in danger of hiding actually important data.

**Comment by technosophos** [comment]
@Joshua-Anderson Maybe that progress() function could be colorized so that it prints out the circle,square,triangle sequence. That might look cool. And maybe we could put that into the colorizer, too. I think that'd be a nice general feature to have.

File: deisctl/backend/backend.go [hunk 1]
```diff
@@ -1,14 +1,25 @@
...
-	Start([]string, *sync.WaitGroup, chan string, chan error)
-	Stop([]string, *sync.WaitGroup, chan string, chan error)
-	Scale(string, int, *sync.WaitGroup, chan string, chan error)
+	/*
+		Create([]string, *sync.WaitGroup, chan string, chan error)
+		Destroy([]string, *sync.WaitGroup, chan string, chan error)
+		Start([]string, *sync.WaitGroup, chan string, chan error)
+		Stop([]string, *sync.WaitGroup, chan string, chan error)
+		Scale(string, int, *sync.WaitGroup, chan string, chan error)
+	*/
```
Comment (mboersma) [inline deisctl/backend/backend.go]: Could you remove this comment? Version control should be sufficient to document the previous interface.

File: deisctl/backend/fleet/start.go [hunk 2]
```diff
@@ -67,18 +69,20 @@ func doStart(c *FleetClient, target string, wg *sync.WaitGroup, outchan chan str
...
-			errchan <- fmt.Errorf("could not find unit: %v", name)
+			fmt.Fprintf(ew, "Could not find unit: %v\n", name)
 			return
 		}
 
 		// if subState changed, send it across the output channel
 		if lastSubState != currentState.SystemdSubState {
-			outchan <- fmt.Sprintf("\033[0;33m%v:\033[0m %v/%v                                 \r",
-				name, currentState.SystemdActiveState, currentState.SystemdSubState)
+			//fmt.Fprintf(out, stateFmt, name, currentState.SystemdActiveState, currentState.SystemdSubState)
```
Comment (mboersma) [inline deisctl/backend/fleet/start.go]: The commented code here should be removed.

File: deisctl/backend/fleet/stop.go [hunk 3]
```diff
@@ -67,18 +71,20 @@ func doStop(c *FleetClient, target string, wg *sync.WaitGroup, outchan chan stri
...
-			errchan <- fmt.Errorf("could not find unit: %v", name)
+			fmt.Fprintf(ew, "Could not find unit: %v\n", name)
 			return
 		}
 
 		// if subState changed, send it across the output channel
 		if lastSubState != currentState.SystemdSubState {
-			outchan <- fmt.Sprintf("\033[0;33m%v:\033[0m %v/%v                                 \r",
-				name, currentState.SystemdActiveState, currentState.SystemdSubState)
+			//fmt.Fprintf(out, stateFmt, name, currentState.SystemdActiveState, currentState.SystemdSubState)
```
Comment (mboersma) [inline deisctl/backend/fleet/stop.go]: The commented code here should be removed.

File: deisctl/cmd/cmd.go [hunk 4]
```diff
@@ -426,6 +401,7 @@ func printState(outchan chan string, errchan chan error, interval time.Duration)
 		time.Sleep(interval)
 	}
 }
+*/
```
Comment (mboersma) [inline deisctl/cmd/cmd.go]: The commented code here should be removed.

**Comment by mboersma** [comment]
This looks good overall, although I don't know where the line overwriting behavior landed. So this probably requires-manual-testing.

**Comment by technosophos** [comment]
The line overwriting stuff was in the commit I merged yesterday. But it shows here (`prettyprint.Overwrite`).

**Comment by technosophos** [comment]
Comments are all cleaned up. Sorry... didn't know about that rule.

Okay, directions for manual testing, where `$DEIS` is the path to the deis source code.

```
$ cd $DEIS/deisctl
$ make test build install
```

Tests should pass and you should then have a shiny new deisctl installed.

Now...

- `deis start SOMETHING` should show you the progress of the start by replacing the text inline.
- `deis help` should correctly display the Deis logo

Edge cases where an error and a message happened in short succession won't stall the app. These were really hard to reproduce, otherwise I would give you a formula. (Note: The hangs on service fail are fixed in #3880, which depends on this PR)

**Comment by mboersma** [comment]
`deisctl help` rules now. But on provisioning it was confusing when the control plane started especially.

1. All "start-pre" statuses overwrote each other on one line, so for a minute it looked like we started only the deis-builder component
2. In the end, I got duplicated status lines, some of which were in the wrong sections:

```
Storage subsystem...
deis-store-monitor.service: active/running                           
deis-store-daemon.service: active/running                            
deis-store-metadata.service: active/running                          
deis-store-gateway@1.service: active/running                         
deis-store-volume.service: active/running                            
Logging subsystem...
deis-logger.service: active/running                                  
deis-logspout.service: active/running                                
Control plane...
deis-controller.service: active/running                              
deis-controller.service: active/running                              
deis-publisher.service: active/running                               
deis-router@1.service: active/running                                
deis-router@2.service: active/running                                
deis-database.service: active/running                                
deis-database.service: active/running                                
deis-router@3.service: active/running                                
deis-registry@1.service: active/running                              
deis-registry@1.service: active/running                              
deis-builder.service: active/running                                 
Data plane...
deis-builder.service: active/running                                 
deis-publisher.service: active/running                               
Routing mesh...
deis-router@2.service: active/running                                
deis-router@3.service: active/running                                
deis-router@1.service: active/running                                
Done.
```

**Comment by technosophos** [comment]
Was that `deisctl start platform`, @mboersma ?

**Comment by mboersma** [comment]
Yes, this was `deisctl start platform` via `SKIP_CLEANUP=true ./tests/bin/test-integration.sh`.

**Comment by technosophos** [comment]
If you run it straight, without a script wrapper, what do you get? I get this:
```
$deisctl stop platform                                                                        
● ▴ ■
■ ● ▴ Stopping Deis...
▴ ■ ●
Routing mesh...
deis-router@1.service: inactive/dead
deis-router@3.service: inactive/dead
deis-router@2.service: inactive/dead
Data plane...
deis-publisher.service: inactive/dead
Control plane...
deis-controller.service: inactive/dead
The service 'deis-builder' failed while stopping.
deis-registry@1.service: inactive/dead
deis-database.service: inactive/dead
Logging subsystem...
deis-logspout.service: inactive/dead
deis-logger.service: inactive/dead
Storage subsystem...
deis-store-gateway@1.service: inactive/dead
The service 'deis-store-volume' failed while stopping.
deis-store-metadata.service: inactive/dead
deis-store-daemon.service: inactive/dead
deis-store-monitor.service: inactive/dead
Done.

Please run `deisctl start platform` to restart Deis.
```

**Comment by mboersma** [comment]
I refetched this branch and rebuilt `deisctl` just now, and `deisctl stop platform` is fine but I still see the duplication issue on `deisctl start platform`:
```console
$ deisctl stop platform
● ▴ ■
■ ● ▴ Stopping Deis...
▴ ■ ●
Routing mesh...
deis-router@1.service: inactive/dead                                 
deis-router@2.service: inactive/dead                                 
deis-router@3.service: inactive/dead                                 
Data plane...
deis-publisher.service: inactive/dead                                
Control plane...
deis-database.service: inactive/dead                                 
deis-registry@1.service: inactive/dead                               
deis-builder.service: inactive/dead                                  
deis-controller.service: inactive/dead                               
Logging subsystem...
deis-logger.service: inactive/dead                                   
deis-logspout.service: inactive/dead                                 
Storage subsystem...
deis-store-gateway@1.service: inactive/dead                          
deis-store-volume.service: inactive/dead                             
deis-store-metadata.service: inactive/dead                           
deis-store-daemon.service: inactive/dead                             
deis-store-monitor.service: inactive/dead                            
Done.
 
Please run `deisctl start platform` to restart Deis.
$ deisctl start platform
● ▴ ■
■ ● ▴ Starting Deis...
▴ ■ ●
Storage subsystem...
deis-store-monitor.service: active/running                           
deis-store-daemon.service: active/running                            
deis-store-metadata.service: active/running                          
deis-store-gateway@1.service: active/running                         
deis-store-volume.service: active/running                            
Logging subsystem...
deis-logger.service: active/running                                  
deis-logspout.service: active/running                                
Control plane...
deis-controller.service: active/running                              
deis-controller.service: active/running                              
deis-database.service: active/running                                
deis-database.service: active/running                                
deis-router@1.service: active/running                                
deis-publisher.service: active/running                               
deis-router@3.service: active/running                                
deis-router@2.service: active/running                                
deis-registry@1.service: active/running                              
deis-registry@1.service: active/running                              
deis-builder.service: activating/start-post
...
```

**Comment by technosophos** [comment]
Okay, the duplicates were removed. This was caused by "background starts" of certain tasks. I switched it to just junk the data coming out of those tasks.

What I see now with `deisctl start` is this:

```
deisctl start platform
● ▴ ■
■ ● ▴ Starting Deis...
▴ ■ ●
Storage subsystem...
deis-store-monitor.service: active/running
deis-store-daemon.service: active/running
deis-store-metadata.service: active/running
deis-store-gateway@1.service: active/running
deis-store-volume.service: active/running
Logging subsystem...
deis-logger.service: active/running
deis-logspout.service: active/running
Control plane...
deis-registry@1.service: active/running
deis-database.service: active/running
deis-controller.service: active/running
deis-builder.service: active/running
Data plane...
deis-publisher.service: active/running
Routing mesh...
deis-router@1.service: active/running
deis-router@2.service: active/running
deis-router@3.service: active/running
Done.
```

**Comment by mboersma** [comment]
LGTM. I tested this again after the most recent changes, and I don't see any duplicated or overlapping output now.

**Comment by carmstrong** [comment]
Code LGTM. :ship:

**Pull request merged**